docs/architecture/adr-063-core-module-api.md

```diff
@@ -0,0 +1,546 @@
+
+The following "core services" are defined by the core API. All valid runtime module implementations should provide
+implementations of these services to modules via both [dependency injection](./adr-057-app-wiring-1.md) and
+manual wiring. The individual services described below are all bundled in a convenient `appmodule.Service`
+"bundle service" so that for simplicity modules can declare a dependency on a single service.
+
+#### Store Services
+
+Store services will be defined in the `cosmossdk.io/core/store` package.
+
+The generic `store.KVStore` interface is the same as current SDK `KVStore` interface. Store keys have been refactored
+into store services which, instead of expecting the context to know about stores, invert the pattern and allow
+retrieving a store from a generic context. There are three store services for the three types of currently supported
+stores - regular kv-store, memory, and transient:
+
+```go
+type KVStoreService interface {
+    OpenKVStore(context.Context) KVStore
+}
+
+type MemoryStoreService interface {
+    OpenMemoryStore(context.Context) KVStore
+}
+type TransientStoreService interface {
+    OpenTransientStore(context.Context) KVStore
+}
+```
+
+Modules can use these services like this:
+```go
+func (k msgServer) Send(ctx context.Context, msg *types.MsgSend) (*types.MsgSendResponse, error) {
+    store := k.kvStoreSvc.OpenKVStore(ctx)
+}
+```
+
+Just as with the current runtime module implementation, modules will not need to explicitly name these store keys,
+but rather the runtime module will choose an appropriate name for them and modules just need to request the
+type of store they need in their dependency injection (or manual) constructors.
+
+#### Event Service
+
+The event `Service` and `Manager` will be defined in the `cosmossdk.io/core/event` package.
+
+The event `Service` gives modules access to an event `Manager` which allows modules to emit typed and legacy,
+untyped events:
+
+```go
+package event
+type Service interface {
+    GetManager(context.Context) Manager
+}
+
+type Manager interface {
+    // Emit emits events to both clients and state machine listeners. These events MUST be emitted deterministically
+    // and should be assumed to be part of blockchain consensus.
+    Emit(proto.Message) error
+    
+    // EmitLegacy emits legacy untyped events to clients only. These events do not need to be emitted deterministically
+    // and are not part of blockchain consensus.
+    EmitLegacy(eventType string, attrs ...LegacyEventAttribute) error
+
+    // EmitClientOnly emits events only to clients. These events do not need to be emitted deterministically
+    // and are not part of blockchain consensus.
+    EmitClientOnly(proto.Message) error
+}
+```
+
+Typed events emitted with `Emit` can be observed by other modules and thus must be deterministic and should be assumed
+to be part of blockchain consensus (whether they are part of the block or app hash is left to the runtime to specify).
```

Comment:
how would runtime do this? would it be up to the application developer?

Reply:
There is a way for registering event listeners described later.